// @validitor/main/public/esm/Types.js
import { Type } from "./Type.js";
function _str() {
    const strType = Type("str", "string");
    strType.setComparason(function (variable) {
        return typeof variable === "string";
    });
    return strType;
}
function _int() {
    const intType = Type("number", "number");
    intType.setComparason(function (variable) {
        return typeof variable === "number";
    });
    return intType;
}
function _bool() {
    const boolType = Type("boolean", "boolean");
    boolType.setComparason(function (variable) {
        return typeof variable === "boolean";
    });
    return boolType;
}
function _bint() {
    const bintType = Type("bignt", "bigint");
    bintType.setComparason(function (variable) {
        return typeof variable === "bigint";
    });
    return bintType;
}
function _obj() {
    const objType = Type("object", "object");
    objType.setComparason(function (variable) {
        return typeof variable === "object";
    });
    return objType;
}
function _func() {
    const funcType = Type("function", "function");
    funcType.setComparason(function (variable) {
        return typeof variable === "function";
    });
    return funcType;
}
function _defined() {
    const definedType = Type("defined", "not undefined");
    definedType.setComparason(function (variable) {
        return variable !== undefined;
    });
    return definedType;
}
function _notdefined() {
    const notdefinedType = Type("notdefined", "undefined");
    notdefinedType.setComparason(function (variable) {
        return variable === undefined;
    });
    return notdefinedType;
}
function _typenull() {
    const nullType = Type("typenull", "null");
    nullType.setComparason(function (variable) {
        return variable === null;
    });
    return nullType;
}
function _any() {
    const anyType = Type("any", "any");
    anyType.setComparason(function (variable) {
        return true;
    });
    return anyType;
}
export const int = _int();
export const any = _any();
export const str = _str();
export const bool = _bool();
export const bint = _bint();
export const obj = _obj();
export const func = _func();
export const defined = _defined();
export const notdefined = _notdefined();
export const typenull = _typenull();
