// @validitor/main/Types.ts
import {Type}from  "./Type.js"; // requiring type class 


function _str(){ // string 
  const strType:any =  Type("str","string"); // setting new type 
  strType.setComparason(function(variable:any):boolean{ // setting the verrification
    return  typeof variable === "string";
  })
  return strType;
}
function _int(){ // string 
  const intType = Type("number","number"); // setting new type 
  intType.setComparason(function(variable:any):boolean{ // setting the verrification
    return  typeof variable === "number";
  })
  return intType;
}

function _bool(){ // boolean 
  const boolType =  Type("boolean","boolean"); // setting new type 
  boolType.setComparason(function(variable:any):boolean{ // setting the verrification
    return  typeof variable === "boolean";
  })
  return boolType;
}
function _bint(){ // bigint 
  const bintType =  Type("bignt","bigint"); // setting new type 
  bintType.setComparason(function(variable:any):boolean{ // setting the verrification
    return typeof variable === "bigint";
  })
  return bintType;
}
function _obj(){ // object 
  const objType =  Type("object","object"); // setting new type 
  objType.setComparason(function(variable:any):boolean{ // setting the verrification
    return  typeof variable === "object";
  })
  return objType;
}
function _func(){ // function 
  const funcType =  Type("function","function"); // setting new type 
  funcType.setComparason(function(variable:any):boolean{ // setting the verrification
    return  typeof variable === "function";
  })
  return funcType;
}
function _defined(){ // !undefined
  const definedType =  Type("defined","not undefined"); // setting new type 
  definedType.setComparason(function(variable:any):boolean{ // setting the verrification
    return  variable !== undefined ;
  })
  return definedType;
}
function _notdefined(){ // undefined
  const notdefinedType = Type("notdefined","undefined"); // setting new type 
  notdefinedType.setComparason (function(variable:any):boolean{ // setting the verrification
    return  variable === undefined;
  })
  return notdefinedType;
}

function _typenull(){ // null
  const nullType =  Type("typenull","null"); // setting new type 
  nullType.setComparason(function(variable:any):boolean{ // setting the verrification
    return  variable === null;
  })
  return nullType;
}
function _any(){ // null
  const anyType =  Type("any","any"); // setting new type 
  anyType.setComparason(function(variable:any):boolean{ // setting the verrification
    return  true;
  })
  return anyType;
}


 export const int = _int();
 export const  any = _any();
 export const str = _str();
 export const bool = _bool();
 export const bint = _bint();
 export const obj = _obj();
 export const func =_func();
 export const defined = _defined();
 export const notdefined = _notdefined();
 export const typenull = _typenull();
