// @validitor/main/public/cjs/main.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Interface = exports.onlyIf = exports.Types = exports.Type = exports.Verify = exports.Validate = void 0;
const Validate_js_1 = require("./Validate.js");
const Verify_js_1 = require("./Verify.js");
const Type_js_1 = require("./Type.js");
const _Types = __importStar(require("./Types.js"));
const onlyIf_js_1 = require("./onlyIf.js");
const Interface_js_1 = require("./Interface.js");
exports.Validate = Validate_js_1.Validate;
exports.Verify = Verify_js_1.Verify;
exports.Type = Type_js_1.Type;
exports.Types = _Types;
exports.onlyIf = onlyIf_js_1.onlyIf;
exports.Interface = Interface_js_1.Interface;
