// @validitor/main/public/cjs/Types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.typenull = exports.notdefined = exports.defined = exports.func = exports.obj = exports.bint = exports.bool = exports.str = exports.any = exports.int = void 0;
const Type_js_1 = require("./Type.js");
function _str() {
    const strType = (0, Type_js_1.Type)("str", "string");
    strType.setComparason(function (variable) {
        return typeof variable === "string";
    });
    return strType;
}
function _int() {
    const intType = (0, Type_js_1.Type)("number", "number");
    intType.setComparason(function (variable) {
        return typeof variable === "number";
    });
    return intType;
}
function _bool() {
    const boolType = (0, Type_js_1.Type)("boolean", "boolean");
    boolType.setComparason(function (variable) {
        return typeof variable === "boolean";
    });
    return boolType;
}
function _bint() {
    const bintType = (0, Type_js_1.Type)("bignt", "bigint");
    bintType.setComparason(function (variable) {
        return typeof variable === "bigint";
    });
    return bintType;
}
function _obj() {
    const objType = (0, Type_js_1.Type)("object", "object");
    objType.setComparason(function (variable) {
        return typeof variable === "object";
    });
    return objType;
}
function _func() {
    const funcType = (0, Type_js_1.Type)("function", "function");
    funcType.setComparason(function (variable) {
        return typeof variable === "function";
    });
    return funcType;
}
function _defined() {
    const definedType = (0, Type_js_1.Type)("defined", "not undefined");
    definedType.setComparason(function (variable) {
        return variable !== undefined;
    });
    return definedType;
}
function _notdefined() {
    const notdefinedType = (0, Type_js_1.Type)("notdefined", "undefined");
    notdefinedType.setComparason(function (variable) {
        return variable === undefined;
    });
    return notdefinedType;
}
function _typenull() {
    const nullType = (0, Type_js_1.Type)("typenull", "null");
    nullType.setComparason(function (variable) {
        return variable === null;
    });
    return nullType;
}
function _any() {
    const anyType = (0, Type_js_1.Type)("any", "any");
    anyType.setComparason(function (variable) {
        return true;
    });
    return anyType;
}
exports.int = _int();
exports.any = _any();
exports.str = _str();
exports.bool = _bool();
exports.bint = _bint();
exports.obj = _obj();
exports.func = _func();
exports.defined = _defined();
exports.notdefined = _notdefined();
exports.typenull = _typenull();
